calicoctl/commands/ipam/check.go

```diff
@@ -0,0 +1,576 @@
+	clusterInfo, err := c.v3Client.ClusterInformation().Get(ctx, "default", options.GetOptions{})
+	if err != nil {
+		return err
+	}
+	c.clusterType = clusterInfo.Spec.ClusterType
+	c.clusterInfoRevision = clusterInfo.ResourceVersion
+	c.datastoreLocked = clusterInfo.Spec.DatastoreReady != nil && !*clusterInfo.Spec.DatastoreReady
+	c.clusterGUID = clusterInfo.Spec.ClusterGUID
+
+	var numAllocs int
+	{
+		fmt.Println("Loading all IPAM blocks...")
+		blocks, err := c.backendClient.List(ctx, model.BlockListOptions{}, "")
+		if err != nil {
+			return fmt.Errorf("failed to list IPAM blocks: %w", err)
+		}
+		fmt.Printf("Found %d IPAM blocks.\n", len(blocks.KVPairs))
+
+		for _, kvp := range blocks.KVPairs {
+			b := kvp.Value.(*model.AllocationBlock)
+			affinity := "<none>"
+			if b.Affinity != nil {
+				affinity = *b.Affinity
+			}
+			fmt.Printf(" IPAM block %s affinity=%s:\n", b.CIDR, affinity)
+			for ord, attrIdx := range b.Allocations {
+				if attrIdx == nil {
+					continue // IP is not allocated
+				}
+				numAllocs++
+				c.recordAllocation(b, ord)
+			}
+		}
+		fmt.Printf("IPAM blocks record %d allocations.\n", numAllocs)
+		fmt.Println()
+	}
+	var activeIPPools []*cnet.IPNet
+	{
+		fmt.Println("Loading all IPAM pools...")
+		ipPools, err := c.v3Client.IPPools().List(ctx, options.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to load IP pools: %w", err)
+		}
+		for _, p := range ipPools.Items {
+			if p.Spec.Disabled {
+				continue
+			}
+			fmt.Printf("  %s\n", p.Spec.CIDR)
+			_, cidr, err := cnet.ParseCIDR(p.Spec.CIDR)
+			if err != nil {
+				return fmt.Errorf("failed to parse IP pool CIDR: %w", err)
+			}
+			activeIPPools = append(activeIPPools, cidr)
+		}
+		fmt.Printf("Found %d active IP pools.\n", len(activeIPPools))
+		fmt.Println()
+	}
+
+	{
+		fmt.Println("Loading all nodes.")
+		nodes, err := c.v3Client.Nodes().List(ctx, options.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to list nodes: %w", err)
+		}
+		numNodeIPs := 0
+		for _, n := range nodes.Items {
+			ips, err := getNodeIPs(n)
+			if err != nil {
+				return err
+			}
+			for _, ip := range ips {
+				c.recordInUseIP(ip, n, fmt.Sprintf("Node(%s)", n.Name))
+				numNodeIPs++
+			}
+		}
+		fmt.Printf("Found %d node tunnel IPs.\n", numNodeIPs)
+		fmt.Println()
+	}
+
+	{
+		fmt.Println("Loading all workload endpoints.")
+		weps, err := c.v3Client.WorkloadEndpoints().List(ctx, options.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to list workload endpoints: %w", err)
+		}
+		numWEPIPs := 0
+		for _, w := range weps.Items {
+			ips, err := getWEPIPs(w)
+			if err != nil {
+				return err
+			}
+			for _, ip := range ips {
+				c.recordInUseIP(ip, w, fmt.Sprintf("Workload(%s/%s)", w.Namespace, w.Name))
+				numWEPIPs++
+			}
+		}
+		fmt.Printf("Found %d workload IPs.\n", numWEPIPs)
+		fmt.Printf("Workloads and nodes are using %d IPs.\n", len(c.inUseIPs))
+		fmt.Println()
+	}
+
+	{
+		const numNodesToPrint = 20
```

Comment:
Should we make this a configurable value?

Reply:
eh, I think it's not super important.